presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -620,6 +639,40 @@ public static long hash(long rawLo, long rawHi)
         return XxHash64.hash(rawLo) ^ XxHash64.hash(unpackUnsignedLong(rawHi));
     }
 
+    public static String toStringDestructive(Slice decimal)
+    {
+        if (isZero(decimal)) {
+            return "0";
+        }
+
+        String[] digitGroup = new String[MAX_PRECISION / MAX_POWER_OF_TEN_INT + 1];
```

Comment:
What's the benefit of doing it int-at-a-time? Integer.toString() has to do a bunch of divisions internally, allocate a char[], a String, etc., and this method needs to maintain the digitGroup array, so I'm not sure it saves much (from a performance point of view) while it significantly complicates the code.

Compare it with:

```java
char[] buffer = new char[MAX_PRECISION + 1];
int index = buffer.length;
do {
    int remainder = divide(decimal, 10, decimal);
    buffer[--index] = (char) ('0' + remainder);
}
while (!isZero(decimal));

if (isNegative(decimal)) {
    buffer[--index] = '-';
}

return new String(buffer, index, buffer.length - index);
```

Unless the performance is significantly better, I'd go with the "obvious" implementation.